File: keycloak/Back/services/authService.js
const axios = require('axios');
const { nanoid } = require('nanoid');
const { User, Role, OTP, Permission, TrustedDevice } = require('../models');
const otpService = require('./otpService');
const { transporter } = require('../config/smtp');
const { sendSMS } = require('../config/sms');
const logger = require('../utils/logger');
require('dotenv').config();

const ERROR_MESSAGES = {
    INVALID_CREDENTIALS: 'Wrong email or password.',
    USER_NOT_FOUND: 'Account not found.',
    KEYCLOAK_MISMATCH: 'Account issue. Contact support.',
    NO_OTP_METHOD: 'No phone or email set for OTP.',
    OTP_SEND_FAILED: 'Couldn’t send OTP. Try again.',
    INVALID_OTP: 'Wrong or expired OTP.',
    INVALID_REFRESH_TOKEN: 'Can’t refresh session. Log in again.',
    PASSWORD_RESET_FAILED: 'Couldn’t start password reset. Try again.',
    PASSWORD_UPDATE_FAILED: 'Couldn’t update password. Try again.',
    KEYCLOAK_UNAVAILABLE: 'Authentication service down. Try again.',
    KEYCLOAK_USER_NOT_FOUND: 'Account not found in system.',
    KEYCLOAK_ADMIN_TOKEN_FAILED: 'Server issue. Try again.',
    DATABASE_ERROR: 'Database issue. Try again.',
    MISSING_FIELDS: 'Missing required fields for 2FA verification.',
    INVALID_KEYCLOAK_RESPONSE: 'Authentication failed. Please try again.',
};

const verificationCache = new Map();
const CACHE_TTL = 5000;

const KEYCLOAK_URL = process.env.KEYCLOAK_URL || 'http://localhost:8080';
const REALM = process.env.REALM || 'TraceFlow';
const CLIENT_ID = process.env.KEYCLOAK_CLIENT_ID || 'traceflow-backend';
const CLIENT_SECRET = process.env.KEYCLOAK_CLIENT_SECRET || 'your-client-secret-from-keycloak';

class AuthService {
    static async getKeycloakAdminToken() {
        try {
            const response = await axios.post(
                `${KEYCLOAK_URL}/realms/master/protocol/openid-connect/token`,
                new URLSearchParams({
                    grant_type: 'password',
                    client_id: 'admin-cli',
                    username: process.env.KEYCLOAK_ADMIN_USER || 'admin',
                    password: process.env.KEYCLOAK_ADMIN_PASSWORD || 'admin',
                })
            );
            return response.data.access_token;
        } catch (error) {
            logger.error(`Keycloak admin token error: ${error.message}`);
            throw new Error(ERROR_MESSAGES.KEYCLOAK_ADMIN_TOKEN_FAILED);
        }
    }

    static async syncKeycloakUser(identifier, keycloakId) {
        try {
            let user = await User.findOne({ where: { email: identifier } });
            if (!user) {
                user = await User.create({
                    userID: `usr_${nanoid()}`,
                    keycloakId,
                    email: identifier,
                    firstname: 'Unknown',
                    lastname: 'User',
                    phone: 'N/A',
                    wallet: `wallet_${nanoid()}`,
                    password: 'KEYCLOAK_MANAGED',
                });
            } else if (!user.keycloakId) {
                await user.update({ keycloakId });
            } else if (user.keycloakId !== keycloakId) {
                throw new Error(ERROR_MESSAGES.KEYCLOAK_MISMATCH);
            }
            return user;
        } catch (error) {
            logger.error(`Sync user error for ${identifier}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.KEYCLOAK_MISMATCH
                ? error
                : new Error(ERROR_MESSAGES.DATABASE_ERROR);
        }
    }

    static async login(identifier, password, deviceIdentifier, otpMethod = 'email', res) {
        let loginResponse;
        try {
            loginResponse = await axios.post(
                `${KEYCLOAK_URL}/realms/${REALM}/protocol/openid-connect/token`,
                new URLSearchParams({
                    grant_type: 'password',
                    client_id: CLIENT_ID,
                    client_secret: CLIENT_SECRET,
                    username: identifier,
                    password,
                    scope: 'openid email profile roles',
                })
            );

            if (
                !loginResponse.data ||
                !loginResponse.data.access_token ||
                !loginResponse.data.refresh_token ||
                !loginResponse.data.expires_in
            ) {
                throw new Error(ERROR_MESSAGES.INVALID_KEYCLOAK_RESPONSE);
            }
        } catch (error) {
            logger.error(`Keycloak login error for ${identifier}: ${error.message}`);
            throw new Error(
                error.message === ERROR_MESSAGES.INVALID_KEYCLOAK_RESPONSE
                    ? error.message
                    : ERROR_MESSAGES.INVALID_CREDENTIALS
            );
        }

        let keycloakUserResponse;
        try {
            const adminToken = await this.getKeycloakAdminToken();
            keycloakUserResponse = await axios.get(
                `${KEYCLOAK_URL}/admin/realms/${REALM}/users?username=${identifier}&exact=true`,
                { headers: { Authorization: `Bearer ${adminToken}` } }
            );
            if (!keycloakUserResponse.data.length) throw new Error(ERROR_MESSAGES.KEYCLOAK_USER_NOT_FOUND);
        } catch (error) {
            logger.error(`Keycloak user fetch error for ${identifier}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.KEYCLOAK_USER_NOT_FOUND
                ? error
                : new Error(ERROR_MESSAGES.KEYCLOAK_UNAVAILABLE);
        }

        const keycloakId = keycloakUserResponse.data[0].id;
        const user = await this.syncKeycloakUser(identifier, keycloakId);

        const userWithDetails = await User.findOne({
            where: { keycloakId },
            include: [
                {
                    model: Role,
                    through: { attributes: [] },
                    include: [{ model: Permission, through: { attributes: [] }, attributes: ['name', 'class'] }],
                },
            ],
        });

        if (!userWithDetails) {
            logger.error(`User details fetch failed for ${identifier}`);
            throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
        }

        const trustedDevice = await TrustedDevice.findOne({
            where: { userID: user.userID, deviceIdentifier, status: 'active' },
        });
        if (trustedDevice) {
            await trustedDevice.update({ lastUsed: new Date() });
            return this.generateLoginResponse(
                userWithDetails,
                loginResponse.data.access_token,
                loginResponse.data.refresh_token,
                loginResponse.data.expires_in,
                res
            );
        }

        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        const phoneRegex = /^\+?\d{8,12}$/;
        const hasValidEmail = user.email && emailRegex.test(user.email);
        const hasValidPhone = user.phone && user.phone !== 'N/A' && phoneRegex.test(user.phone);

        if (!hasValidEmail && !hasValidPhone) {
            logger.error(`No valid OTP method for ${user.userID}`);
            throw new Error(ERROR_MESSAGES.NO_OTP_METHOD);
        }

        try {
            let otp;
            let selectedMethod = otpMethod;

            if (selectedMethod === 'email' && hasValidEmail) {
                try {
                    otp = await otpService.generateOTP(user.userID, 'user');
                    await transporter.sendMail({
                        from: process.env.SMTP_USER,
                        to: user.email,
                        subject: 'TraceFlow 2FA OTP',
                        text: `Your OTP is ${otp.code}. It expires in 10 minutes.`,
                    });
                } catch (error) {
                    logger.error(`Email OTP send failed for ${user.userID}: ${error.message}`);
                    if (hasValidPhone) {
                        selectedMethod = 'phone';
                    } else {
                        throw new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
                    }
                }
            }

            if (!otp && selectedMethod === 'phone' && hasValidPhone) {
                try {
                    otp = await otpService.generateOTP(user.userID, 'user');
                    const smsResult = await sendSMS(user.phone, `Your TraceFlow OTP is ${otp.code}`, 'otp');
                    if (!smsResult.success) {
                        logger.error(`SMS OTP send failed for ${user.userID}: ${smsResult.reason}`);
                        if (hasValidEmail) {
                            selectedMethod = 'email';
                            otp = await otpService.generateOTP(user.userID, 'user');
                            await transporter.sendMail({
                                from: process.env.SMTP_USER,
                                to: user.email,
                                subject: 'TraceFlow 2FA OTP',
                                text: `Your OTP is ${otp.code}. It expires in 10 minutes.`,
                            });
                        } else {
                            throw new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
                        }
                    } else {
                        selectedMethod = smsResult.fallback ? 'email' : 'phone';
                    }
                } catch (error) {
                    logger.error(`SMS OTP send failed for ${user.userID}: ${error.message}`);
                    throw new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
                }
            }

            if (!otp) {
                logger.error(`OTP delivery failed for ${user.userID}`);
                throw new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
            }

            return {
                requires2FA: true,
                userID: user.userID,
                deviceIdentifier,
                tempToken: loginResponse.data.access_token,
                refreshToken: loginResponse.data.refresh_token,
                expiresIn: loginResponse.data.expires_in,
                message: `OTP sent to your ${selectedMethod}`,
            };
        } catch (error) {
            logger.error(`OTP error for ${identifier}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.NO_OTP_METHOD
                ? error
                : new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
        }
    }

    static async verify2FA(userID, otpCode, deviceIdentifier, trustDevice, tempToken, refreshToken, res) {
        const missingFields = [];
        if (!userID) missingFields.push('userID');
        if (!otpCode) missingFields.push('otpCode');
        if (!deviceIdentifier) missingFields.push('deviceIdentifier');
        if (trustDevice === undefined) missingFields.push('trustDevice');
        if (!tempToken) missingFields.push('tempToken');
        if (!refreshToken) missingFields.push('refreshToken');

        if (missingFields.length > 0) {
            throw new Error(`${ERROR_MESSAGES.MISSING_FIELDS} Missing: ${missingFields.join(', ')}`);
        }

        const cacheKey = `${userID}:${otpCode}:${deviceIdentifier}`;
        const cachedResult = verificationCache.get(cacheKey);
        if (cachedResult) {
            return cachedResult;
        }

        const user = await User.findByPk(userID);
        if (!user) {
            logger.error(`2FA verification failed: User ${userID} not found`);
            throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
        }

        try {
            await otpService.validateOTP(user.userID, otpCode, 'user');
        } catch (error) {
            logger.error(`2FA OTP validation failed for ${userID}: ${error.message}`);
            throw new Error(ERROR_MESSAGES.INVALID_OTP);
        }

        const userWithDetails = await User.findOne({
            where: { userID },
            include: [
                {
                    model: Role,
                    through: { attributes: [] },
                    include: [{ model: Permission, through: { attributes: [] }, attributes: ['name', 'class'] }],
                },
            ],
        });

        if (trustDevice) {
            const existingDevice = await TrustedDevice.findOne({
                where: { userID, deviceIdentifier },
            });
            if (existingDevice) {
                await existingDevice.update({ status: 'active', lastUsed: new Date() });
            } else {
                await TrustedDevice.create({
                    userID,
                    deviceIdentifier,
                    status: 'active',
                    lastUsed: new Date(),
                    expiresAt: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000),
                });
            }
        }

        const result = this.generateLoginResponse(userWithDetails, tempToken, refreshToken, 900, res);
        verificationCache.set(cacheKey, result);
        setTimeout(() => verificationCache.delete(cacheKey), CACHE_TTL);

        return result;
    }

    static async refreshToken(refreshToken, res) {
        try {
            const response = await axios.post(
                `${KEYCLOAK_URL}/realms/${REALM}/protocol/openid-connect/token`,
                new URLSearchParams({
                    grant_type: 'refresh_token',
                    client_id: CLIENT_ID,
                    client_secret: CLIENT_SECRET,
                    refresh_token: refreshToken,
                })
            );

            if (
                !response.data ||
                !response.data.access_token ||
                !response.data.refresh_token ||
                !response.data.expires_in
            ) {
                throw new Error(ERROR_MESSAGES.INVALID_REFRESH_TOKEN);
            }

            const cookieOptions = {
                httpOnly: true,
                secure: process.env.NODE_ENV === 'production',
                sameSite: 'Lax',
                path: '/',
            };

            res.cookie('accessToken', response.data.access_token, {
                ...cookieOptions,
                maxAge: response.data.expires_in * 1000,
            });
            res.cookie('refreshToken', response.data.refresh_token, {
                ...cookieOptions,
                maxAge: 7 * 24 * 60 * 60 * 1000,
            });

            return {
                user: { message: 'Token refreshed' },
                accessToken: response.data.access_token,
                refreshToken: response.data.refresh_token,
            };
        } catch (error) {
            logger.error(`Token refresh error: ${error.message}`);
            throw new Error(ERROR_MESSAGES.INVALID_REFRESH_TOKEN);
        }
    }

    static generateLoginResponse(user, token, refreshToken, expiresIn, res) {
        if (!user || !token || !refreshToken || !expiresIn) {
            logger.error(`Invalid login response data for user ${user?.userID || 'unknown'}`);
            throw new Error(ERROR_MESSAGES.INVALID_KEYCLOAK_RESPONSE);
        }

        const roles = user.Roles?.map((role) => ({
            roleID: role.roleID,
            name: role.name,
            description: role.description || undefined,
            permissions: role.Permissions ? role.Permissions.map((p) => ({
                permissionID: p.permissionID,
                name: p.name,
                class: p.class,
                description: p.description || undefined,
            })) : [],
        })) || [];

        const cookieOptions = {
            httpOnly: true,
            secure: false,
            maxAge: expiresIn * 1000,
            sameSite: 'Lax',
            path: '/',
        };

        res.cookie('accessToken', token, cookieOptions);
        res.cookie('refreshToken', refreshToken, {
            ...cookieOptions,
            maxAge: 7 * 24 * 60 * 60 * 1000,
        });

        return {
            requires2FA: false,
            accessToken: token,
            user: {
                userID: user.userID,
                email: user.email,
                phone: user.phone,
                roles,
            },
        };
    }

    static async resend2FA(userID, otpMethod = 'email') {
        const user = await User.findByPk(userID);
        if (!user) {
            logger.error(`Resend 2FA failed: User ${userID} not found`);
            throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
        }

        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        const phoneRegex = /^\+?\d{8,12}$/;
        const hasValidEmail = user.email && emailRegex.test(user.email);
        const hasValidPhone = user.phone && user.phone !== 'N/A' && phoneRegex.test(user.phone);

        if (!hasValidEmail && !hasValidPhone) {
            logger.error(`No valid OTP method for ${userID}`);
            throw new Error(ERROR_MESSAGES.NO_OTP_METHOD);
        }

        try {
            let otp;
            let selectedMethod = otpMethod;

            if (selectedMethod === 'email' && hasValidEmail) {
                otp = await otpService.generateOTP(userID, 'user');
                await transporter.sendMail({
                    from: process.env.SMTP_USER,
                    to: user.email,
                    subject: 'TraceFlow OTP',
                    text: `Your OTP is ${otp.code}. It expires in 10 minutes.`,
                });
            } else if (selectedMethod === 'phone' && hasValidPhone) {
                otp = await otpService.generateOTP(userID, 'user');
                const smsResult = await sendSMS(user.phone, `Your TraceFlow OTP is ${otp.code}`, 'otp');
                if (!smsResult.success) {
                    logger.error(`SMS resend failed for ${userID}: ${smsResult.reason}`);
                    if (hasValidEmail) {
                        selectedMethod = 'email';
                        otp = await otpService.generateOTP(userID, 'user');
                        await transporter.sendMail({
                            from: process.env.SMTP_USER,
                            to: user.email,
                            subject: 'TraceFlow OTP',
                            text: `Your OTP is ${otp.code}. It expires in 10 minutes.`,
                        });
                    } else {
                        throw new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
                    }
                } else {
                    selectedMethod = smsResult.fallback ? 'email' : 'phone';
                }
            } else {
                logger.error(`Invalid OTP method for ${userID}: ${otpMethod}`);
                throw new Error(ERROR_MESSAGES.NO_OTP_METHOD);
            }

            return { userID, message: `OTP resent to your ${selectedMethod}` };
        } catch (error) {
            logger.error(`Resend OTP error for ${userID}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.NO_OTP_METHOD
                ? error
                : new Error(ERROR_MESSAGES.OTP_SEND_FAILED);
        }
    }

    static async initiatePasswordReset(identifier) {
        let keycloakId, user;
        try {
            const adminToken = await this.getKeycloakAdminToken();
            const userResponse = await axios.get(
                `${KEYCLOAK_URL}/admin/realms/${REALM}/users?username=${identifier}&exact=true`,
                { headers: { Authorization: `Bearer ${adminToken}` } }
            );
            if (!userResponse.data.length) throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
            keycloakId = userResponse.data[0].id;
            user = await this.syncKeycloakUser(identifier, keycloakId);
        } catch (error) {
            logger.error(`Password reset init error for ${identifier}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.USER_NOT_FOUND
                ? error
                : new Error(ERROR_MESSAGES.KEYCLOAK_UNAVAILABLE);
        }

        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        const phoneRegex = /^\+?\d{8,12}$/;
        const hasValidEmail = user.email && emailRegex.test(user.email);
        const hasValidPhone = user.phone && user.phone !== 'N/A' && phoneRegex.test(user.phone);

        try {
            const otp = await otpService.generateOTP(user.userID, 'user');
            let selectedMethod = 'email';

            if (hasValidEmail) {
                await transporter.sendMail({
                    from: process.env.SMTP_USER,
                    to: user.email,
                    subject: 'TraceFlow Password Reset OTP',
                    text: `Your OTP for password reset is ${otp.code}. It expires in 10 minutes.`,
                });
            } else if (hasValidPhone) {
                const smsResult = await sendSMS(user.phone, `Your TraceFlow password reset OTP is ${otp.code}`, 'otp');
                if (!smsResult.success) {
                    logger.error(`SMS reset failed for ${user.userID}: ${smsResult.reason}`);
                    throw new Error(ERROR_MESSAGES.PASSWORD_RESET_FAILED);
                }
                selectedMethod = smsResult.fallback ? 'email' : 'phone';
            } else {
                logger.error(`No valid OTP method for ${user.userID}`);
                throw new Error(ERROR_MESSAGES.NO_OTP_METHOD);
            }

            return { userID: user.userID, message: `OTP sent to your ${selectedMethod}` };
        } catch (error) {
            logger.error(`Password reset OTP error for ${user.userID}: ${error.message}`);
            throw error.message === ERROR_MESSAGES.NO_OTP_METHOD
                ? error
                : new Error(ERROR_MESSAGES.PASSWORD_RESET_FAILED);
        }
    }

    static async verifyPasswordResetOTP(userID, otpCode) {
        const user = await User.findByPk(userID);
        if (!user) {
            logger.error(`Password reset OTP verification failed: User ${userID} not found`);
            throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
        }

        try {
            await otpService.validateOTP(user.userID, otpCode, 'user');
        } catch (error) {
            logger.error(`Password reset OTP validation failed for ${userID}: ${error.message}`);
            throw new Error(ERROR_MESSAGES.INVALID_OTP);
        }

        const tempToken = nanoid();
        await User.update({ tempResetToken: tempToken }, { where: { userID } });

        return { userID, tempToken, message: 'OTP verified. Proceed to reset password.' };
    }

    static async resetPassword(userID, newPassword, tempToken) {
        const user = await User.findByPk(userID);
        if (!user) {
            logger.error(`Password reset failed: User ${userID} not found`);
            throw new Error(ERROR_MESSAGES.USER_NOT_FOUND);
        }
        if (user.tempResetToken !== tempToken) {
            logger.error(`Password reset failed for ${userID}: Invalid reset token`);
            throw new Error('Invalid or expired reset token');
        }

        try {
            const adminToken = await this.getKeycloakAdminToken();
            await axios.put(
                `${KEYCLOAK_URL}/admin/realms/${REALM}/users/${user.keycloakId}/reset-password`,
                { type: 'password', value: newPassword, temporary: false },
                { headers: { Authorization: `Bearer ${adminToken}` } }
            );
            await User.update({ tempResetToken: null }, { where: { userID } });
        } catch (error) {
            logger.error(`Password reset error for ${userID}: ${error.message}`);
            throw new Error(ERROR_MESSAGES.PASSWORD_UPDATE_FAILED);
        }

        return { message: 'Password reset successfully' };
    }
}

module.exports = AuthService;

File: Code/New folder/mobile/lib/services/auth_service.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_web_auth_2/flutter_web_auth_2.dart';
import 'package:jwt_decoder/jwt_decoder.dart';
import '../utils/constants.dart';
import './cookie_manager.dart';
import './http_client.dart';

class AuthService {
  static Future<Map<String, dynamic>> initiateKeycloakLogin() async {
    try {
      final authUrl =
          '$keycloakUrl/realms/$realm/protocol/openid-connect/auth?client_id=$clientId&redirect_uri=${Uri.encodeComponent(redirectUri)}&response_type=code&scope=openid%20email%20profile&kc_idp_hint=google&prompt=select_account';
      final result = await FlutterWebAuth2.authenticate(
        url: authUrl,
        callbackUrlScheme: 'http',
      );
      final code = Uri.parse(result).queryParameters['code'];
      if (code == null) throw Exception('No authorization code received');

      final tokenResponse = await http.post(
        Uri.parse('$keycloakUrl/realms/$realm/protocol/openid-connect/token'),
        headers: {'Content-Type': 'application/x-www-form-urlencoded'},
        body: {
          'grant_type': 'authorization_code',
          'code': code,
          'redirect_uri': redirectUri,
          'client_id': clientId,
          'client_secret': clientSecret,
        },
      );
      if (tokenResponse.statusCode != 200) {
        throw Exception(
          'Failed to exchange code: ${json.decode(tokenResponse.body)['error_description'] ?? tokenResponse.body}',
        );
      }

      final tokens = json.decode(tokenResponse.body);
      final accessToken = tokens['access_token'] as String;
      final refreshToken = tokens['refresh_token'] as String;
      await CookieManager.saveCookies({
        'accessToken': accessToken,
        'refreshToken': refreshToken,
      });

      final userResponse = await CustomHttpClient.get(
        Uri.parse('$baseUrl/test'),
      );
      if (userResponse.statusCode != 200) {
        throw Exception('Failed to fetch user data: ${userResponse.body}');
      }

      final userData = json.decode(userResponse.body);
      _ensureRoles(userData, accessToken);

      final requires2FA = await _check2FARequired();
      if (requires2FA) {
        return {
          'userID': userData['user']?['userID']?.toString(),
          'requires2FA': true,
          'tempToken': accessToken,
          'refreshToken': refreshToken,
          'expiresIn': tokens['expires_in'] * 1000,
          'otpMethod': 'phone',
        };
      }

      return {
        'accessToken': accessToken,
        'refreshToken': refreshToken,
        'expiresIn': tokens['expires_in'] * 1000,
        'user': userData['user'],
      };
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<bool> _check2FARequired() async {
    try {
      final response = await CustomHttpClient.get(
        Uri.parse('$baseUrl/auth/check-2fa'),
      );
      if (response.statusCode == 200) {
        final result = json.decode(response.body);
        return result['requires2FA'] as bool? ?? false;
      }
      throw Exception('Failed to check 2FA requirement');
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> checkAuthStatus() async {
    try {
      final response = await CustomHttpClient.get(Uri.parse('$baseUrl/test'));
      if (response.statusCode == 200) {
        final result = json.decode(response.body);
        final accessToken = CookieManager.cookies['accessToken'];
        if (accessToken != null) _ensureRoles(result, accessToken);
        return result;
      }
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> login(
      String identifier,
      String password,
      String deviceIdentifier,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/login'),
        body: json.encode({
          'identifier': identifier,
          'password': password,
          'deviceIdentifier': deviceIdentifier,
          'otpMethod': 'phone',
        }),
      );
      if (response.statusCode == 200) {
        final result = json.decode(response.body);
        if (kDebugMode) print('Login response: ${jsonEncode(result)}');
        if (result['accessToken'] != null) {
          _ensureRoles(result, result['accessToken']);
          await CookieManager.saveCookies({
            'accessToken': result['accessToken'],
            'refreshToken': result['refreshToken'] ?? '',
          });
        }
        return result;
      }
      throw Exception(_parseError(response));
    } catch (e) {
      if (kDebugMode) print('Login error: $e');
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> verify2FA(
      String userID,
      String otpCode,
      bool trustDevice,
      String tempToken,
      String refreshToken,
      String deviceIdentifier,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/2fa/verify'),
        body: json.encode({
          'userID': userID,
          'otpCode': otpCode,
          'trustDevice': trustDevice,
          'tempToken': tempToken,
          'refreshToken': refreshToken,
          'deviceIdentifier': deviceIdentifier,
        }),
      );
      if (response.statusCode == 200) {
        final result = json.decode(response.body);
        if (result['accessToken'] != null) {
          _ensureRoles(result, result['accessToken']);
          await CookieManager.saveCookies({
            'accessToken': result['accessToken'],
            'refreshToken': result['refreshToken'] ?? '',
          });
        }
        return result;
      }
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> refreshToken(String refreshToken) async {
    try {
      // Ensure refreshToken is stored in cookies before the request
      await CookieManager.saveCookies({'refreshToken': refreshToken});
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/refresh'),
        // No body or custom Content-Type needed; backend reads from cookies
      );
      if (response.statusCode == 200) {
        final result = json.decode(response.body);
        await CookieManager.saveCookies({
          'accessToken': result['accessToken'],
          'refreshToken': result['refreshToken'] ?? refreshToken,
        });
        return result;
      }
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }
  static Future<Map<String, dynamic>> resend2FA(
      String userID,
      String otpMethod,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/2fa/resend'),
        body: json.encode({'userID': userID, 'otpMethod': otpMethod}),
      );
      if (response.statusCode == 200) return json.decode(response.body);
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> initiatePasswordReset(
      String identifier,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/password/reset/initiate'),
        body: json.encode({'identifier': identifier}),
      );
      if (response.statusCode == 200) return json.decode(response.body);
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<Map<String, dynamic>> verifyPasswordResetOTP(
      String userID,
      String otpCode,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/password/reset/verify'),
        body: json.encode({'userID': userID, 'otpCode': otpCode}),
      );
      if (response.statusCode == 200) return json.decode(response.body);
      throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<void> resetPassword(
      String userID,
      String newPassword,
      String tempToken,
      ) async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/password/reset'),
        body: json.encode({
          'userID': userID,
          'newPassword': newPassword,
          'tempToken': tempToken,
        }),
      );
      if (response.statusCode != 200) throw Exception(_parseError(response));
    } catch (e) {
      throw Exception(_parseError(e));
    }
  }

  static Future<void> logout() async {
    try {
      final response = await CustomHttpClient.post(
        Uri.parse('$baseUrl/auth/logout'),
      );
      if (response.statusCode != 200) throw Exception(_parseError(response));
    } catch (e) {
      if (kDebugMode) print('Logout error: $e');
    } finally {
      await CookieManager.clearCookies();
    }
  }

  static void _ensureRoles(Map<String, dynamic> data, String accessToken) {
    if (data['user'] == null ||
        data['user']['roles'] == null ||
        (data['user']['roles'] as List).isEmpty) {
      final decodedToken = JwtDecoder.decode(accessToken);
      final realmRoles =
          (decodedToken['realm_access']?['roles'] as List<dynamic>?) ?? [];
      data['user'] ??= {'roles': []};
      data['user']['roles'] = realmRoles
          .where((role) => role != null)
          .toList()
          .asMap()
          .entries
          .map(
            (e) => {
          'roleID': (e.key + 1).toString(),
          'name': e.value.toString(),
          'description': null,
          'permissions': [],
        },
      )
          .toList();
    }
  }

  static String _parseError(dynamic input) {
    if (input is http.Response) {
      try {
        final body = json.decode(input.body);
        return body['error'] ??
            body['message'] ??
            body['error_description'] ??
            _getErrorMessage(input.statusCode);
      } catch (_) {
        return _getErrorMessage(input.statusCode);
      }
    }
    return input.toString().replaceFirst('Exception: ', '');
  }

  static String _getErrorMessage(int statusCode) {
    switch (statusCode) {
      case 400:
        return 'Invalid request. Please check your input.';
      case 401:
        return 'Authentication failed. Please check your credentials.';
      case 403:
        return 'Account locked or access denied.';
      case 404:
        return 'Account not found.';
      case 429:
        return 'Too many attempts. Please wait.';
      case 500:
        return 'Server error. Please try again later.';
      default:
        return 'An unexpected error occurred.';
    }
  }

  static Future<dynamic> makeAuthenticatedRequest({
    required Future<http.Response> Function() request,
  }) async {
    try {
      final response = await request();
      if (response.statusCode == 401) {
        final refreshToken = CookieManager.cookies['refreshToken'];
        if (refreshToken == null || refreshToken.isEmpty) {
          throw Exception('No refresh token available');
        }
        try {
          final refreshResult = await refreshToken;
          if (kDebugMode) print('Refresh result: $refreshResult');
          final retryResponse = await request();
          return json.decode(retryResponse.body);
        } catch (e) {
          if (e.toString().contains('Invalid refresh token')) {
            await CookieManager.clearCookies();
          }
          throw Exception('Authentication failed');
        }
      }
      return json.decode(response.body);
    } catch (e) {
      if (kDebugMode) print('Authenticated request failed: $e');
      rethrow;
    }
  }
}